### PR TITLE
rustdoc: collect trait impls as an early pass

### LINKED ISSUE #36922: rustdoc: Trait implementations aren't listed in generated docs under some circumstances
Serde's `Deserialize` and `Serialize` traits don't show up in the "Trait Implementations" section for types that use custom derive through Macros 1.1 or through Syntex. [Alex Crichton also notes](https://github.com/rust-lang/rust/issues/35900#issuecomment-251039515) that rustdoc does not display `Copy` in the "Trait Implementations" section with the following code:

``` rust
#[derive(Clone)]           
pub struct Point {         
    x: i32,                
    y: i32,                
}                          

const _FOO: () = {         
    impl Copy for Point {} 
    ()                     
};
```

I've only tested this on nightly Rust.

### LINKED ISSUE #41480: Rustdoc doesn't see trait implementations inside of functions
It's possible to define trait implementations inside of functions in Rust. This is useful not too often, but particularly nice if you have a macro that both defines trait implementation and calls some initialization for a given type.

Anyway, the following code is valid in Rust itself:

```rust
pub struct S;

pub fn a() {
    impl From<u8> for S {
        fn from(_: u8) -> S {
            S
        }
    }
}
```

But Rustdoc doesn't generate documentation for a given trait, despite it being visible to Rust itself, and being immediately available and unique across entire crate.

I can move `impl` out of the function and then it gets seen by Rustdoc, but this is rather inconvenient when `impl` is generated as part of a bigger macro.

### LINKED ISSUE #52545: Rustdoc ignores traits implemented inside of a function
After #51952 landed, I attempted to make Diesel compatible with recent nightlies by following the advice of wrapping the output in a function rather than a module. However, we are now unable to build our documentation. The compiler complains that every derived trait within Diesel is not implemented. This only occurs when running `cargo doc`, the crate otherwise compiles fine. I have tested with 1.26.1, stable, and the most recent nightly.

### PR DESCRIPTION
**Pull request by QuietMisdreavus**
Fixes https://github.com/rust-lang/rust/issues/52545, fixes https://github.com/rust-lang/rust/issues/41480, fixes https://github.com/rust-lang/rust/issues/36922

Right now, rustdoc pulls all its impl information by scanning a crate's HIR for any items it finds. However, it doesn't recurse into anything other than modules, preventing it from seeing trait impls that may be inside things like functions or consts. Thanks to https://github.com/rust-lang/rust/pull/53002, now these items actually *exist* for rustdoc to see, but they still weren't getting collected for display.

But there was a secret. Whenever we pull in an item from another crate, we don't have any of its impls in the local HIR, so instead we ask the compiler for *everything* and filter out after the fact. This process is only triggered if there's a cross-crate re-export in the crate being documented, which can sometimes leave this info out of the docs. This PR instead moves this collection into an early pass, which occurs immediately after crate cleaning, so that that collection occurs regardless. In addition, by including the HIR's own `trait_impls` in addition to the existing `all_trait_implementations` calls, we can collect all these tricky trait impls without having to scan for them!

### CODE REVIEW

**Comment by sgrif** [comment]
❤️ ❌ 💯

**Comment by QuietMisdreavus** [comment]
Ohhhhhh boy, i broke a lot of things. Back to the drawing board...

**Comment by GuillaumeGomez** [comment]
Can't wait for you to finish this! :)

**Comment by QuietMisdreavus** [comment]
Pushed a new commit that (at least locally) should fix the issue that caused linkchecker to fail. Let's see if this works...

A possible refactor that i'd like to do, but which may be out of scope for this PR by now, is to move the trait impls to a different list instead of piling them into the root of the crate. This echoes what the HIR already does, since that's the list we're using to extract local impls here.

File: src/librustdoc/clean/inline.rs [hunk 1]
```diff
@@ -371,30 +308,60 @@ pub fn build_impl(cx: &DocContext, did: DefId, ret: &mut Vec<clean::Item>) {
...
     let predicates = tcx.predicates_of(did);
-    let trait_items = tcx.associated_items(did).filter_map(|item| {
-        if associated_trait.is_some() || item.vis == ty::Visibility::Public {
-            Some(item.clean(cx))
-        } else {
-            None
+    let (trait_items, generics) = if let Some(nodeid) = tcx.hir.as_local_node_id(did) {
+        match tcx.hir.expect_item(nodeid).node {
+            hir::ItemKind::Impl(.., ref gen, _, _, ref item_ids) => {
+                (
```
Comment (GuillaumeGomez) [inline src/librustdoc/clean/inline.rs]: Super weird indent, didn't understand at first what I was reading haha.

File: src/librustdoc/clean/inline.rs [hunk 2]
```diff
@@ -575,6 +548,10 @@ fn separate_supertrait_bounds(mut g: clean::Generics)
 }
 
 pub fn record_extern_trait(cx: &DocContext, did: DefId) {
+    if did.is_local() {
+        return;
```
Comment (GuillaumeGomez) [inline src/librustdoc/clean/inline.rs]: You're not constant about your `return` calls: some have `;`, some others don't. ;)

File: src/librustdoc/passes/mod.rs [hunk 3]
```diff
@@ -296,18 +305,22 @@ impl<'a> fold::DocFolder for ImplStripper<'a> {
...
+                    debug!("ImplStripper: impl item for stripped trait; removing");
                     return None;
                 }
             }
             if let Some(generics) = imp.trait_.as_ref().and_then(|t| t.generics()) {
                 for typaram in generics {
                     if let Some(did) = typaram.def_id() {
                         if did.is_local() && !self.retained.contains(&did) {
+                            debug!("ImplStripper: stripped item in trait's generics; \
+                                   removing impl");
```
Comment (GuillaumeGomez) [inline src/librustdoc/passes/mod.rs]: Bad indent.

File: src/test/rustdoc/doc-proc-macro.rs [hunk 4]
```diff
@@ -8,6 +8,8 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+// ignore-stage1
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/doc-proc-macro.rs]: 👍 👍 👍 👍

File: src/test/rustdoc/inline_local/trait-vis.rs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+// Copyright 2012-2013 The Rust Project Developers. See the COPYRIGHT
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/inline_local/trait-vis.rs]: I think date can be changed.

File: src/librustdoc/clean/inline.rs (hunk 1 above)
Comment (QuietMisdreavus) [reply src/librustdoc/clean/inline.rs]: Is there a good way to do something like this, though? The first item of the tuple is too long to just fit on one line.

File: src/librustdoc/clean/inline.rs (hunk 2 above)
Comment (QuietMisdreavus) [reply src/librustdoc/clean/inline.rs]: The `return` lines without semicolons are ones i didn't write myself. :P

File: src/librustdoc/passes/mod.rs (hunk 3 above)
Comment (QuietMisdreavus) [reply src/librustdoc/passes/mod.rs]: What's the proper way to indent this, then? The string is too long to fit on one line.

File: src/test/rustdoc/inline_local/trait-vis.rs (hunk 5 above)
Comment (QuietMisdreavus) [reply src/test/rustdoc/inline_local/trait-vis.rs]: Oops, that's what i get for copying headers blindly. :sweat_smile:

File: src/librustdoc/clean/inline.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/inline.rs]: No, it's fine as is. Just wanted to comment on it though.

File: src/librustdoc/clean/inline.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/inline.rs]: The diff doesn't agree. :p

File: src/librustdoc/passes/mod.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/passes/mod.rs]: The content of the string should be after the `"`:

```
"long
 string"

// and not:
"long
string"
```

**Comment by GuillaumeGomez** [comment]
With the tests it's now all good for me. r=me once CI is green.

**Comment by QuietMisdreavus** [comment]
@ollie27 Because it seemed like a good way to structure it. I wanted it to happen after the initial cleaning, but before items were stripped. Over time i want to try to move out more of the "additional processing" steps into passes, so that we can splice out more of the code that's currently sitting around `clean/mod.rs`.

**Comment by QuietMisdreavus** [comment]
Failure is legit. Investigating...

**Comment by ollie27** [comment]
The following ICEs in the same way when using `--document-private-items`:

```rust
pub struct Foo;

fn f() {
    trait Bar {}
    impl Bar for Foo {}
}
```

We may need to strip impls for traits defined inside functions even when using `--document-private-items`.

---

I've also noticed that this PR doesn't fix the issue for inherent impls like the following:

```rust
pub struct Foo;

fn f() {
    impl Foo {
        pub fn bar() {}
    }
}
````

`bar` is still missing from the docs.

**Comment by QuietMisdreavus** [comment]
Aha, that looks promising! I guess we need to make sure we strip out impls for traits that aren't public.

Re: inherent impls, i was under the impression that those functions wouldn't be available outside the function scope, but [i was totally wrong](https://play.rust-lang.org/?gist=b72f36976646f8d557d4a737cb655b88&version=stable&mode=debug&edition=2015) i need to rethink a few things. This PR won't grab those without adding more steps - since it's not a trait impl, it won't be in the `trait_impls` map in the HIR. It'll have to be grabbed in a follow-up PR.

**Comment by QuietMisdreavus** [comment]
I was able to get past that ICE, but then i hit another one:

```
 Documenting rustc_codegen_llvm v0.0.0 (file:///home/misdreavus/git/rust-builds/src/librustc_codegen_llvm)                                             error[E0275]: overflow evaluating the requirement `std::boxed::Box<std::vec::Vec<syntax::ast::Attribute>>: std::marker::Sync`                            |                                                                                                                                                      = help: consider adding a `#![recursion_limit="128"]` attribute to your crate                                                                          = note: required because it appears within the type `std::option::Option<std::boxed::Box<std::vec::Vec<syntax::ast::Attribute>>>`                      = note: required because it appears within the type `syntax::ThinVec<syntax::ast::Attribute>`                                                          = note: required because it appears within the type `syntax::ast::FieldPat`
  = note: required because it appears within the type `syntax::source_map::Spanned<syntax::ast::FieldPat>`
  = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<syntax::source_map::Spanned<syntax::ast::FieldP
at>>`                                                                                                                                                    = note: required because it appears within the type `alloc::raw_vec::RawVec<syntax::source_map::Spanned<syntax::ast::FieldPat>>`
  = note: required because it appears within the type `std::vec::Vec<syntax::source_map::Spanned<syntax::ast::FieldPat>>`                                = note: required because it appears within the type `syntax::ast::PatKind`                                                                             = note: required because it appears within the type `syntax::ast::Pat`                                                                                 = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<syntax::ast::Pat>`                               = note: required because it appears within the type `std::boxed::Box<syntax::ast::Pat>`
  = note: required because it appears within the type `syntax::ptr::P<syntax::ast::Pat>`
  = note: required because it appears within the type `syntax::ast::Local`
  = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<syntax::ast::Local>`
  = note: required because it appears within the type `std::boxed::Box<syntax::ast::Local>`
  = note: required because it appears within the type `syntax::ptr::P<syntax::ast::Local>`
  = note: required because it appears within the type `syntax::ast::StmtKind`
  = note: required because it appears within the type `syntax::ast::Stmt`
  = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<syntax::ast::Stmt>`                              = note: required because it appears within the type `alloc::raw_vec::RawVec<syntax::ast::Stmt>`                                                        = note: required because it appears within the type `std::vec::Vec<syntax::ast::Stmt>`                                                                 = note: required because it appears within the type `syntax::ast::Block`
  = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<syntax::ast::Block>`
  = note: required because it appears within the type `std::boxed::Box<syntax::ast::Block>`
  = note: required because it appears within the type `syntax::ptr::P<syntax::ast::Block>`                                                               = note: required because it appears within the type `syntax::ast::ExprKind`
  = note: required because it appears within the type `syntax::ast::Expr`                                                                                = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<syntax::ast::Expr>`                              = note: required because it appears within the type `std::boxed::Box<syntax::ast::Expr>`                                                               = note: required because it appears within the type `syntax::ptr::P<syntax::ast::Expr>`                                                                = note: required because it appears within the type `syntax::ast::AnonConst`
  = note: required because it appears within the type `syntax::ast::TyKind`
  = note: required because it appears within the type `syntax::ast::Ty`
  = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<syntax::ast::Ty>`
  = note: required because it appears within the type `std::boxed::Box<syntax::ast::Ty>`
  = note: required because it appears within the type `syntax::ptr::P<syntax::ast::Ty>`
  = note: required because it appears within the type `syntax::ast::GenericArg`
  = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<syntax::ast::GenericArg>`
  = note: required because it appears within the type `alloc::raw_vec::RawVec<syntax::ast::GenericArg>`
  = note: required because it appears within the type `std::vec::Vec<syntax::ast::GenericArg>`
  = note: required because it appears within the type `syntax::ast::AngleBracketedArgs`
  = note: required because it appears within the type `syntax::ast::GenericArgs`
  = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<syntax::ast::GenericArgs>`
  = note: required because it appears within the type `std::boxed::Box<syntax::ast::GenericArgs>`
  = note: required because it appears within the type `syntax::ptr::P<syntax::ast::GenericArgs>`
  = note: required because it appears within the type `std::option::Option<syntax::ptr::P<syntax::ast::GenericArgs>>`
  = note: required because it appears within the type `syntax::ast::PathSegment`
  = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<syntax::ast::PathSegment>`
  = note: required because it appears within the type `alloc::raw_vec::RawVec<syntax::ast::PathSegment>`
  = note: required because it appears within the type `std::vec::Vec<syntax::ast::PathSegment>`
  = note: required because it appears within the type `syntax::ast::Path`
  = note: required because it appears within the type `syntax::ast::MetaItem`
  = note: required because it appears within the type `std::option::Option<syntax::ast::MetaItem>`
  = note: required because it appears within the type `rustc::middle::cstore::NativeLibrary`
  = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<rustc::middle::cstore::NativeLibrary>`
  = note: required because it appears within the type `alloc::raw_vec::RawVec<rustc::middle::cstore::NativeLibrary>`
  = note: required because it appears within the type `std::vec::Vec<rustc::middle::cstore::NativeLibrary>`
  = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<std::vec::Vec<rustc::middle::cstore::NativeLibrar
y>>`
  = note: required because it appears within the type `(rustc::hir::def_id::CrateNum, std::sync::Arc<std::vec::Vec<rustc::middle::cstore::NativeLibrary
>>)`
  = note: required because it appears within the type `std::marker::PhantomData<(rustc::hir::def_id::CrateNum, std::sync::Arc<std::vec::Vec<rustc::midd
le::cstore::NativeLibrary>>)>`
  = note: required because it appears within the type `std::collections::hash::table::RawTable<rustc::hir::def_id::CrateNum, std::sync::Arc<std::vec::V
ec<rustc::middle::cstore::NativeLibrary>>>`
  = note: required because it appears within the type `std::collections::HashMap<rustc::hir::def_id::CrateNum, std::sync::Arc<std::vec::Vec<rustc::midd
le::cstore::NativeLibrary>>, std::hash::BuildHasherDefault<rustc_data_structures::fx::FxHasher>>`
  = note: required because it appears within the type `CrateInfo`
  = note: required because it appears within the type `CodegenResults`

error: Could not document `rustc_codegen_llvm`.
```

Several crates in the compiler docs will generate similar errors, where trying to generate the auto-trait impls for types that somehow contain an AST type will fail because the type tree is too deep. This works on master, so i'll have to figure out what i changed to mess this up. Investigating...

**Comment by QuietMisdreavus** [comment]
I've amended the last commit and fixed travis.

**Comment by GuillaumeGomez** [comment]
Let's retry!

@bors: r+

**Comment by QuietMisdreavus** [comment]
Ah, someone bumped `parking_lot` in the lockfile. I'll adjust it for rustdoc here and retry, but i want to make sure it still builds locally first.

**Comment by QuietMisdreavus** [comment]
Merge conflict resolved.

@bors r=GuillaumeGomez

**Comment by kennytm** [comment]
@bors p=30

Rollup fairness

**Comment by QuietMisdreavus** [comment]
Lockfile merge conflict again. Rebased, will re-approve once travis is green.

**Comment by GuillaumeGomez** [comment]
Some day it'll get merged!

**Comment by ollie27** [comment]
```
[01:06:38] failures:
[01:06:38]     [rustdoc] rustdoc/intra-link-extern-crate.rs
[01:06:38]     [rustdoc] rustdoc/intra-link-private.rs
[01:06:38]     [rustdoc] rustdoc/issue-29503.rs
```

The failure in `rustdoc/issue-29503.rs` is because #53689 wasn't fixed properly, #54087 fixes it.

The other failures are from intra-links added in #51885 which don't work.

**Comment by QuietMisdreavus** [comment]
Oh wait, i see what's happening on the other two. The impls pulled in from `all_trait_implementations` are getting scanned for intra-links, and those links aren't valid. I'll need to perform "impls on non-documented types/traits stripping" before intra-doc links are collected, so that these false positives don't hit.

I also want to change how intra-doc links are collected to pull the "parent scope" ID out differently, so that impls made inside functions still resolve correctly. (And so that i don't have to try to fit impls into their correct module!)

**Comment by QuietMisdreavus** [comment]
(*Once more, with feeling...*)

I've rebased to include https://github.com/rust-lang/rust/pull/54087, which fixes one of the tests.

I've also included two new commits:

* The first is a fix for the test failures. It turns out that by including a bunch of extraneous new trait impls into the crate data before collecting intra-doc links, they were getting scanned for links. As libstd is starting to include more and more intra-doc links into its docs, some of these impls brought some along. This was causing some false-positives for resolution failures. As such, i scan impls being pulled in from other crates to make sure they're actually relevant to the current crate before exiting the pass. Crate-local impls are always loaded.
* The second adds a new test for a bug i realized i introduced with this PR. As far as the language (and most of rustdoc) is concerned, trait impls don't care about where they're located, so they can be dumped anywhere. (That was the entire impetus for this PR, if you recall.) However, there *is* one place where the location of the impl can matter: intra-doc links. So i patched intra-doc links to get the HIR module parent instead of just using the location in the cleaned crate as the resolution scope.
  * External items should still be able to use their location in the cleaned crate for their resolution scope. (Yeah, this means that impls pulled in from other crates that actually survive that pass are getting resolved at the crate root. `>_>`)
  * There is a major wrinkle to this, which i had to leave behind: Our current implementation uses `Resolver::with_scope` to use a specific scope for intra-links. However, this function can only use *modules* as the given scope. I couldn't find any related methods for constraining to an arbitrary block scope, so as of this PR intra-doc links on trait impl blocks use their module scope instead of their local scope for resolution. Since i grab the module parent for everything else, trait impls directly in modules should be able to use the usual resolution scope for intra-links.

**Comment by GuillaumeGomez** [comment]
Let's go again! :D

@bors: r+

**Comment by QuietMisdreavus** [comment]
*lockfiles* 😠 

Going to rebase again, will force-push and watch travis afterward.

**Comment by GuillaumeGomez** [comment]
Conflict again. :-/ Just r=me once rebased.

**Comment by QuietMisdreavus** [comment]
🔥 😡 ***lockfiles*** 😡 🔥

Going to wait for travis again to make sure i didn't accidentally break the build. The next couple PRs in the queue don't modify the lockfile, so with that last priority bump this should hopefully not have to deal with this again?

**Pull request merged**